pkg/globalnet/controllers/ingress_pod_controller.go

```diff
@@ -0,0 +1,123 @@
+	"github.com/submariner-io/admiral/pkg/util"
+	submarinerv1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+)
+
+func startIngressPodController(svc *corev1.Service, config syncer.ResourceSyncerConfig) (*ingressPodController, error) {
+	var err error
+
+	_, gvr, err := util.ToUnstructuredResource(&submarinerv1.GlobalIngressIP{}, config.RestMapper)
+	if err != nil {
+		return nil, err
+	}
+
+	controller := &ingressPodController{
+		baseSyncerController: newBaseSyncerController(),
+		ingressIPs:           config.SourceClient.Resource(*gvr),
+		scheme:               config.Scheme,
+		svcName:              svc.Name,
+		namespace:            svc.Namespace,
+		ingressIPMap:         stringset.NewSynchronized(),
+	}
+
+	labelSelector := labels.Set(svc.Spec.Selector).AsSelector().String()
+
+	controller.resourceSyncer, err = syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:                "Ingress Pod syncer",
+		ResourceType:        &corev1.Pod{},
+		SourceClient:        config.SourceClient,
+		SourceNamespace:     svc.Namespace,
+		RestMapper:          config.RestMapper,
+		Federator:           federate.NewCreateFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll),
+		Scheme:              config.Scheme,
+		Transform:           controller.process,
+		SourceLabelSelector: labelSelector,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := controller.Start(); err != nil {
+		return nil, err
+	}
+
+	klog.Infof("Created Pod controller for (%s/%s) with selector %q", svc.Namespace, svc.Name, labelSelector)
+
+	return controller, nil
+}
+
+func (c *ingressPodController) process(from runtime.Object, numRequeues int, op syncer.Operation) (runtime.Object, bool) {
+	pod := from.(*corev1.Pod)
+
+	ingressIP := &submarinerv1.GlobalIngressIP{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      getIngressIPName(pod),
+			Namespace: pod.Namespace,
+		},
+	}
+
+	if op == syncer.Delete {
+		c.ingressIPMap.Remove(ingressIP.Name)
+		klog.Infof("ingress Pod %s/%s deleted", pod.Namespace, pod.Name)
+
+		return ingressIP, false
+	}
+
+	if c.ingressIPMap.Contains(ingressIP.Name) || pod.Status.Phase != corev1.PodRunning || pod.Status.PodIP == "" {
+		// Avoid assigning ingressIPs to pods that are not ready with an endpoint IP
+		return nil, false
+	}
+
+	klog.Infof("%q ingress Pod %s/%s", op, pod.Namespace, pod.Name)
+
+	ingressIP.Spec = submarinerv1.GlobalIngressIPSpec{
+		Target:     submarinerv1.HeadlessServicePod,
+		ServiceRef: &corev1.LocalObjectReference{Name: c.svcName},
+		PodRef:     &corev1.LocalObjectReference{Name: pod.Name},
+	}
+
+	c.ingressIPMap.Add(ingressIP.Name)
+
+	return ingressIP, false
+}
+
+func (c *ingressPodController) CleanupIngressIPs() {
+	// TODO: Use DeleteCollection()
```

Comment:
There is a small window where this could get called prior to the ingressIPMap being (fully) populated. You could use the resource syncer cache (`GetResources`) which is fully populated on startup.